### PR TITLE
Data loss bugs with SPLITSHARDs and RELOAD

### PR DESCRIPTION
**Pull request by chatman**
Here are fixes for the following:
https://issues.apache.org/jira/browse/SOLR-13718
https://issues.apache.org/jira/browse/SOLR-12291

### CODE REVIEW

**Comment by dragonsinth** [comment]
Can we be sure to land these as individual commits on master?

File: solr/CHANGES.txt [hunk 1]
```diff
@@ -28,11 +50,21 @@ Velocity 1.7 and Velocity Tools 2.0
...
+  to them being included outside those keys for backwards compatibility). See SOLR-12708 for more 
+  details
+
 Bug Fixes
 ----------------------
 
 * SOLR-13349: High CPU usage in Solr due to Java 8 bug (Erick Erickson)
 
+* SOLR-12708: Async collection actions should not hide internal failures (Mano Kovacs, Varun Thacker, Tomás Fernández Löbbe)
+
```
[the remark this answers is not in the document]
Comment (chatman) [reply solr/CHANGES.txt]: Okay, sure. Will add them in FULLSTORY_CHANGES.txt.

**Comment by chatman** [comment]
> Can we be sure to land these as individual commits on master?

Master of upstream Solr? They're already there:
https://github.com/apache/lucene-solr/commits/branch_7_7
https://github.com/apache/lucene-solr/commits/master

Fullstorydev's lucene-solr master? That's still on 6x. I can find out if these issues affect 6x and backport, if you suggest so.

**Comment by dragonsinth** [comment]
> > Can we be sure to land these as individual commits on master?
> 
> Fullstorydev's lucene-solr master? That's still on 6x. I can find out if these issues affect 6x and backport, if you suggest so.

Oops, I misspoke, not master.  I meant "Let's be sure to push individual commits onto `solr7_7_1` rather than merging this PR as a merge or squash commit."

**Comment by chatman** [comment]
> > > Can we be sure to land these as individual commits on master?
> > 
> > 
> > Fullstorydev's lucene-solr master? That's still on 6x. I can find out if these issues affect 6x and backport, if you suggest so.
> 
> Oops, I misspoke, not master. I meant "Let's be sure to push individual commits onto `solr7_7_1` rather than merging this PR as a merge or squash commit."

Ah, got you! Sure, I'll do a regular merge (not a squashed one).

**Comment by dragonsinth** [comment]
> Ah, got you! Sure, I'll do a regular merge (not a squashed one).

Don't do that either.  LOL. Make sure it ends up being a fast-forward that includes 4 commits.

**Comment by dragonsinth** [comment]
Um.. you still wanted to get LGTMs from @noblepaul @jbdutton @EmilyBragg  :D

**Comment by chatman** [comment]
Ouch, sorry for jumping the gun! I'll wait for their review.

**Comment by chatman** [comment]
And here's one more problem of my jumping the gun: https://github.com/fullstorydev/lucene-solr/pull/36

File: solr/core/src/java/org/apache/solr/cloud/api/collections/DeleteCollectionCmd.java [hunk 2]
```diff
@@ -106,17 +105,13 @@ public void call(ClusterState state, ZkNodeProps message, NamedList results) thr
...
-      }
 
       Set<String> okayExceptions = new HashSet<>(1);
       okayExceptions.add(NonExistentCoreException.class.getName());
 
-      List<Replica> failedReplicas = ocmh.collectionCmd(message, params, results, null, asyncId, requestMap, okayExceptions);
-      for (Replica failedRepilca : failedReplicas) {
-        boolean isSharedFS = failedRepilca.getBool(ZkStateReader.SHARED_STORAGE_PROP, false) && failedRepilca.get("dataDir") != null;
+      List<Replica> failedReplicas = ocmh.collectionCmd(message, params, results, null, asyncId, okayExceptions);
+      for (Replica failedReplica : failedReplicas) {
```
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/DeleteCollectionCmd.java]: 💯 for making it easier to search code :)

File: solr/core/src/java/org/apache/solr/cloud/api/collections/RestoreCmd.java [hunk 3]
```diff
@@ -241,9 +245,12 @@ public void call(ClusterState state, ZkNodeProps message, NamedList results) thr
       Assign.AssignStrategy assignStrategy = assignStrategyFactory.create(clusterState, restoreCollection);
       List<ReplicaPosition> replicaPositions = assignStrategy.assign(ocmh.cloudManager, assignRequest);
       sessionWrapper = PolicyHelper.getLastSessionWrapper(true);
+
+      CountDownLatch countDownLatch = new CountDownLatch(restoreCollection.getSlices().size());
+
       //Create one replica per shard and copy backed up data to it
       for (Slice slice : restoreCollection.getSlices()) {
-        log.debug("Adding replica for shard={} collection={} ", slice.getName(), restoreCollection);
+        log.info("Adding replica for shard={} collection={} ", slice.getName(), restoreCollection);
```
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/RestoreCmd.java]: Out of curiosity, why the level change here?

File: solr/core/src/java/org/apache/solr/cloud/api/collections/RestoreCmd.java [hunk 4]
```diff
@@ -274,42 +281,78 @@ public void call(ClusterState state, ZkNodeProps message, NamedList results) thr
...
-          Replica.State stateRep  = r.getState();
+        for (Slice s : restoreCollection.getSlices()) {
+          for (Replica r : s.getReplicas()) {
+            String nodeName = r.getNodeName();
+            String coreNodeName = r.getCoreName();
+            Replica.State stateRep = r.getState();
 
-          log.debug("Calling REQUESTAPPLYUPDATES on: nodeName={}, coreNodeName={}, state={}"
-              , nodeName, coreNodeName, stateRep.name());
+            log.debug("Calling REQUESTAPPLYUPDATES on: nodeName={}, coreNodeName={}, state={}", nodeName, coreNodeName,
```
Comment (EmilyBragg) [inline solr/core/src/java/org/apache/solr/cloud/api/collections/RestoreCmd.java]: ultraminor nit, but can the line break be earlier in the args (to keep things from overflowing the line as well as to make it more obvious that the text on 344 is an arg)?

File: solr/core/src/test/org/apache/solr/cloud/api/collections/AbstractCloudBackupRestoreTestCase.java [hunk 5]
```diff
@@ -312,7 +379,7 @@ private void testBackupAndRestore(String collectionName, int backupReplFactor) t
     if (random().nextBoolean()) {
       assertEquals(0, restore.process(client).getStatus());
     } else {
-      assertEquals(RequestStatusState.COMPLETED, restore.processAndWait(client, 30));//async
+      assertEquals(RequestStatusState.COMPLETED, restore.processAndWait(client, 60));//async
```
Comment (EmilyBragg) [inline solr/core/src/test/org/apache/solr/cloud/api/collections/AbstractCloudBackupRestoreTestCase.java]: What's the reason for this change?

File: solr/core/src/test/org/apache/solr/cloud/api/collections/TestRequestStatusCollectionAPI.java [hunk 6]
```diff
@@ -103,13 +110,18 @@ public void test() throws Exception {
...
     try {
-      message = sendStatusRequestWithRetry(params, MAX_WAIT_TIMEOUT_SECONDS);
+      splitResponse = sendStatusRequestWithRetry(params, MAX_WAIT_TIMEOUT_SECONDS);
+      message = (String) splitResponse.findRecursive("status","msg");
     } catch (SolrServerException | IOException e) {
       e.printStackTrace();
     }
 
     assertEquals("found [1001] in completed tasks", message);
+    // create * 2 + preprecovery *2 + split + req_apply_upd * 2 =7
```
Comment (EmilyBragg) [inline solr/core/src/test/org/apache/solr/cloud/api/collections/TestRequestStatusCollectionAPI.java]: preprecovery?

Comment (EmilyBragg) [reply solr/core/src/test/org/apache/solr/cloud/api/collections/TestRequestStatusCollectionAPI.java]: (asking because I can't really find a description in the docs and just want to know what it's purpose is :) )
